hummingbot/strategy/pure_market_making/pure_market_making_v2.pyx

```diff
@@ -0,0 +1,709 @@
+        self._in_flight_cancels[order_id] = self._current_timestamp
+        market.c_cancel(market_info.symbol, order_id)
+
+    cdef c_start(self, Clock clock, double timestamp):
+        StrategyBase.c_start(self, clock, timestamp)
+        self._last_timestamp = timestamp
+
+    cdef c_tick(self, double timestamp):
+        StrategyBase.c_tick(self, timestamp)
+
+        cdef:
+            int64_t current_tick = <int64_t>(timestamp // self._status_report_interval)
+            int64_t last_tick = <int64_t>(self._last_timestamp // self._status_report_interval)
+            bint should_report_warnings = ((current_tick > last_tick) and
+                                           (self._logging_options & self.OPTION_LOG_STATUS_REPORT))
+            list active_maker_orders = self.active_maker_orders
+
+        try:
+            if not self._all_markets_ready:
+                self._all_markets_ready = all([market.ready for market in self._markets])
+                if not self._all_markets_ready:
+                    # Markets not ready yet. Don't do anything.
+                    if should_report_warnings:
+                        self.logger().warning(f"Markets are not ready. No market making trades are permitted.")
+                    return
+
+            if should_report_warnings:
+                if not all([market.network_status is NetworkStatus.CONNECTED for market in self._markets]):
+                    self.logger().warning(f"WARNING: Some markets are not connected or are down at the moment. Market "
+                                          f"making may be dangerous when markets or networks are unstable.")
+
+            market_info_to_active_orders = self.market_info_to_active_orders
+
+            for market_info in self._market_infos.values():
+                self._delegate_lock = True
+                orders_proposal = None
+                try:
+                    orders_proposal = self.c_get_orders_proposal_for_market_info(
+                        market_info,
+                        market_info_to_active_orders[market_info]
+                    )
+                except Exception:
+                    self.logger().error("Unknown error while generating order proposals.", exc_info=True)
+                finally:
+                    self._delegate_lock = False
+                self.c_execute_orders_proposal(market_info, orders_proposal)
+
+            self.c_check_and_cleanup_shadow_records()
+        finally:
+            self._last_timestamp = timestamp
+
+    cdef object c_get_orders_proposal_for_market_info(self, object market_info, list active_orders):
+        cdef:
+            double last_trade_price
+            MarketBase maker_market = market_info.market
+            int actions = 0
+            list cancel_order_ids = []
+
+        # Before doing anything, ask the filter delegate whether to proceed or not.
+        if not self._filter_delegate.c_should_proceed_with_processing(self, market_info, active_orders):
```

Comment:
As exceptions in filter delegate, return True, shouldn't this be checking the opposite condition. Like if there is an exception in the filter delegate, then NO_Op_Orders is proposed

Reply:
No, if there's an exception from filter delegate, then the exception is raised. The `True?` doesn't mean you'd see a `True` being returned from Python's side. What it means is that the C++ code would return a `True` and set an exception. Cython would automatically check for the exception when it sees a True and raise the exception on Python's side. Thus, you won't see the return value when an exception is raised.